### PR TITLE
feat(@formatjs/cli): A CLI for formatjs

### PR DESCRIPTION
**Pull request by pyrocat101**
Related feature backlog item: formatjs/react-intl/issues/1392 and #68.

---

This PR introduces a CLI for react-intl that is that called `react-intl`. The only subcommand is support is `extract`. Right now this is just a CLI interface to the `babel-plugin-react-intl`. Notably, it also supports TypeScript by applying the correct set of Babel plugin to the files of `.ts` or `.tsx` extension.

# Example

```shell
# Extract messages from all JavaScript and TypeScript under src/ directory 
# to JSON files in messages/ directory.
$ react-cli extract src/**/*.{js,ts,tsx} --messages-dir messages
```

# Options

All CLI options are passed through to the Babel plugins.

```shell
react-intl extract --help
Usage: react-intl extract [options] [files...]

Extract string messages from React components that use react-intl.
The input language is expected to be TypeScript or ES2017 with JSX.

Options:
  --messages-dir <dir>                                  The target location where the plugin will output a `.json` file corresponding to each
                                                        component from which React Intl messages were extracted. If not provided, the extracted
                                                        message descriptors will be printed to standard output.
  --enforce-descriptions                                Whether message declarations _must_ contain a `description` to provide context to
                                                        translators. (default: false)
  --enforce-default-message                             Whether message declarations _must_ contain a `defaultMessage` to fallback when
                                                        translations are not present. (default: true)
  --no-enforce-default-message
  --extract-source-location                             Whether the metadata about the location of the message in the source file should be
                                                        extracted. If `true`, then `file`, `start`, and `end` fields will exist for each extracted
                                                        message descriptors. (default: false)
  --module-source-name <name>                           The ES6 module source name of the React Intl package. Defaults to: `"react-intl"`, but can
                                                        be changed to another name/path to React Intl.
  --remove-default-message                              Remove `defaultMessage` field in generated js after extraction (default: false)
  --additional-component-names <comma-separated-names>  Additional component names to extract messages from, e.g: `['FormattedFooBarMessage']`.
                                                        **NOTE**: By default we check for the fact that `FormattedMessage` & `FormattedHTMLMessage`
                                                        are imported from `moduleSourceName` to make sure variable alias works. This option does
                                                        not do that so it's less safe.
  --extract-from-format-message-call                    Opt-in to extract from `intl.formatMessage` call with the same restrictions, e.g: has to be
                                                        called with object literal such as `intl.formatMessage({ id: 'foo', defaultMessage: 'bar',
                                                        description: 'baz'})` (default: false)
  -h, --help                                            output usage information
```

# Caveat

It uses `@babel/plugin-transform-typescript` to parse TypeScript. So it is subject to all [caveats](https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats) of the parser.

# What is Left?

- [ ] More TypeScript testing
- [x] Write README

### CODE REVIEW

File: packages/babel-plugin-react-intl/package.json [hunk 1]
```diff
@@ -31,5 +33,8 @@
     "url": "https://github.com/formatjs/formatjs/issues"
   },
   "homepage": "https://github.com/formatjs/formatjs#readme",
-  "gitHead": "8b0baec8eda5002715cf893274fe59782fc2d371"
+  "gitHead": "8b0baec8eda5002715cf893274fe59782fc2d371",
```
Comment (pyrocat101) [inline packages/babel-plugin-react-intl/package.json:36]: @longlho How is this generated?

Comment (longlho) [reply packages/babel-plugin-react-intl/package.json:36]: that's a lerna publish artifact, you can actually just delete it

**Comment by longlho** [comment]
can u rename the pkg to `formatjs-cli` instead?

**Review by longlho — APPROVED** [review]

**Comment by montezume** [comment]
Probably a stupid question, but I can't get this to go recursively into all subdirectories under `src.`

**Comment by longlho** [comment]
What args did you pass in?

**Comment by montezume** [comment]
`react-cli extract src/**/*.js --out-file messages.json`

**Comment by longlho** [comment]
hmm and you tried the latest version from npm?

**Comment by greg-hoarau** [comment]
Same for me with : 

`formatjs extract --extract-from-format-message-call ./src/**/*.{js,ts,tsx}`

Result :

`Error: ENOENT: no such file or directory, open './src/**/*.js'
    at Object.openSync (fs.js:440:3)
    at Object.readFileSync (fs.js:342:35)
    at Object.transformFileSync (/home/ghoarau/.nvm/versions/node/v13.1.0/lib/node_modules/@formatjs/cli/node_modules/@babel/core/lib/transform-file.js:82:61)
    at extractSingleFile (/home/ghoarau/.nvm/versions/node/v13.1.0/lib/node_modules/@formatjs/cli/lib/extract.js:61:18)
    at Object.extract [as default] (/home/ghoarau/.nvm/versions/node/v13.1.0/lib/node_modules/@formatjs/cli/lib/extract.js:88:33)
    at Command.<anonymous> (/home/ghoarau/.nvm/versions/node/v13.1.0/lib/node_modules/@formatjs/cli/lib/cli.js:70:32)
    at Command.listener (/home/ghoarau/.nvm/versions/node/v13.1.0/lib/node_modules/@formatjs/cli/node_modules/commander/index.js:359:8)
    at Command.emit (events.js:210:5)
    at Command.parseArgs (/home/ghoarau/.nvm/versions/node/v13.1.0/lib/node_modules/@formatjs/cli/node_modules/commander/index.js:791:12)
    at Command.parse (/home/ghoarau/.nvm/versions/node/v13.1.0/lib/node_modules/@formatjs/cli/node_modules/commander/index.js:562:21)`

Any idea ?

My workaround for now is to do like so : 

``formatjs extract --extract-from-format-message-call `find ./src/ -regex '.*\.\(js\|ts\|tsx\)'` ``

**Comment by longlho** [comment]
thanks for reporting ill take a look asap

**Comment by pyrocat101** [comment]
@montezume Which shell do you use to run the command?

@AlashaHonore Same questions as above. And do you have any `.js` files in your folder? If not, then the wildcard matching will fail. Consider trying only `./src/**/*.{ts,tsx}` if you only want to match against `.ts` and `.tsx` files.

**Comment by greg-hoarau** [comment]
@pyrocat101 I use Ubuntu on Windows Subsystem for Linux (v2). 

I only have `.ts` and `.tsx` files. It fails even with `./src/**/*.{ts,tsx}`

Result : `Error: ENOENT: no such file or directory, open './src/**/*.tsx'`

**Comment by pyrocat101** [comment]
@AlashaHonore if you try `echo ./src/**/*.{ts,tsx}` command, does it still complain? If so, it is mostly likely a wildcard matching problem with the shell you use.

**Comment by montezume** [comment]
@pyrocat101 bash on Mac OSX 10.14

**Comment by greg-hoarau** [comment]
@pyrocat101 

> @AlashaHonore if you try `echo ./src/**/*.{ts,tsx}` command, does it still complain? If so, it is mostly likely a wildcard matching problem with the shell you use.

There is no error when I try `echo ./src/**/*.{ts,tsx}` but : 

`./src/dux/actionTypes.ts ./src/dux/configureStore.ts ./src/dux/rootReducer.ts [... more path with .ts extension ...] ./src/**/*.tsx`

As you can see, `./src/**/*.tsx` is in the result... I think this is because no file matching `*.tsx`. 

I have some `.tsx` files but they're  in a much more deep level. (ex : `./src/modules/modname/comp.tsx`)

The `/**/` match only one subdirectory and not all... Is this the expected behavior ?

**Comment by lostfictions** [comment]
Hey, just bumped into this issue -- at a glance it looks like the cli doesn't actually use [glob](https://github.com/isaacs/node-glob) or any other path-matching library and is instead relying on system globbing (which amongst other things, doesn't support `**` -- that's not a standard shell expansion feature -- and won't work outright on some shells or Windows).

@pyrocat101 can you confirm whether this is the case? The cli probably needs a globbing library to do its job properly.

**Comment by longlho** [comment]
@lostfictions would you mind making a GH issue for easier discussion? Thanks! Looks like using `glob` is the correct solution.

**Pull request merged**